File: content_tool/gemini/fake.py
import json
from typing import Any

from content_tool.gemini.client import GeminiResult


class FakeGeminiClient:
    def __init__(
        self,
        canned_responses: dict[str, dict[str, Any]],
        canned_grounding: dict[str, list[dict[str, Any]]] | None = None,
    ) -> None:
        self._canned = canned_responses
        # Optional per-agent grounding chunks (e.g. for topic_existing_search,
        # which harvests result.grounding_chunks rather than the parsed JSON).
        self._canned_grounding = canned_grounding or {}
        self.calls: list[dict[str, Any]] = []

    def set_audit_response(self, canned: dict[str, Any]) -> None:
        """Set/replace the canned response returned for agent="audit".

        Fills in `severity_summary` if the caller omits it so the returned
        payload validates against AuditOutput.
        """
        payload = dict(canned)
        findings = payload.get("findings", [])
        if "severity_summary" not in payload:
            payload["severity_summary"] = {
                "high": sum(1 for f in findings if f.get("severity") == "high"),
                "medium": sum(1 for f in findings if f.get("severity") == "medium"),
                "low": sum(1 for f in findings if f.get("severity") == "low"),
            }
        if "overall_pass" not in payload:
            payload["overall_pass"] = payload["severity_summary"]["high"] == 0
        self._canned["audit"] = payload

    async def generate(
        self,
        *,
        agent: str,
        system_prompt: str,
        user_prompt: str,
        response_schema: dict[str, Any] | None,
        tools: list[str],
    ) -> GeminiResult:
        self.calls.append({
            "agent": agent,
            "system_prompt": system_prompt,
            "user_prompt": user_prompt,
            "tools": tools,
        })
        # A plain-text call (no schema) need not have a canned JSON body — e.g.
        # topic_existing_search, where only the grounding chunks matter.
        if agent not in self._canned and response_schema is not None:
            raise KeyError(f"No canned response for agent={agent}")
        parsed = self._canned.get(agent, {})
        return GeminiResult(
            parsed=parsed,
            raw_text=json.dumps(parsed, ensure_ascii=False),
            tokens_in=1000,
            tokens_out=500,
            thinking_tokens=100,
            latency_ms=10,
            grounding_chunks=self._canned_grounding.get(agent),
        )

File: deploy/cloudflare-workers/src/gemini/fake.ts
/**
 * Fake Gemini client for tests.
 *
 * Ported from `content_tool/gemini/fake.py`. Returns canned responses keyed by
 * `agent` and records every call for assertions.
 */

import type { GeminiClient, GeminiResult, GenerateOptions } from "./types";

const FAKE_TOKENS_IN = 1000;
const FAKE_TOKENS_OUT = 500;
const FAKE_THINKING_TOKENS = 100;
const FAKE_LATENCY_MS = 10;

/** A recorded `generate` call (mirrors the Python `calls` list). */
export interface RecordedCall {
  agent: string;
  systemPrompt: string;
  userPrompt: string;
  tools: string[];
}

export class FakeGeminiClient implements GeminiClient {
  private readonly canned: Record<string, Record<string, unknown>>;
  // Optional per-agent grounding chunks (e.g. for topic_existing_search, which
  // harvests groundingChunks rather than the parsed JSON).
  private readonly cannedGrounding: Record<string, Record<string, unknown>[]>;
  readonly calls: RecordedCall[] = [];

  constructor(
    cannedResponses: Record<string, Record<string, unknown>>,
    cannedGrounding: Record<string, Record<string, unknown>[]> = {},
  ) {
    this.canned = { ...cannedResponses };
    this.cannedGrounding = { ...cannedGrounding };
  }

  async generate(opts: GenerateOptions): Promise<GeminiResult> {
    const { agent, systemPrompt, userPrompt, responseSchema, tools } = opts;

    this.calls.push({ agent, systemPrompt, userPrompt, tools });

    const canned = this.canned[agent];
    // A plain-text call (no schema) need not have a canned JSON body — e.g.
    // topic_existing_search, where only the grounding chunks matter.
    if (canned === undefined && responseSchema !== null) {
      throw new Error(`No canned response for agent=${agent}`);
    }
    const parsed = canned ?? {};

    return {
      parsed,
      rawText: JSON.stringify(parsed),
      tokensIn: FAKE_TOKENS_IN,
      tokensOut: FAKE_TOKENS_OUT,
      thinkingTokens: FAKE_THINKING_TOKENS,
      latencyMs: FAKE_LATENCY_MS,
      groundingChunks: this.cannedGrounding[agent] ?? null,
      finishReason: null,
    };
  }
}
